dolosse/run_control/cli/kafka_pyinq.py

```diff
@@ -0,0 +1,246 @@
+"""
+file: midas_kafka_control_producer
+brief: command line interface using PyInquirer for producing control commands to kafka
+author: Caswell Pieters
+date: 24 March 2020
+"""
+
+from __future__ import print_function, unicode_literals
+from datetime import datetime
+import json
+import os
+from pathlib import Path
+import regex
+import sys
+import threading
+import yaml
+
+from colorama import Fore, Style
+from confluent_kafka import Producer, Consumer, KafkaException
+from PyInquirer import style_from_dict, Token, prompt, Validator, ValidationError
+
+
+class RunNumberValidator(Validator):
+    def validate(self, document):
+        ok = regex.match('^(0|[1-9][0-9]{0,9})$', document.text)
+        if not ok:
+            raise ValidationError(
+                message='Please enter a positive or zero run number',
+                cursor_position=len(document.text))  # Move cursor to end
+
+
+class PathValidator(Validator):
+    def validate(self, document):
+        ok = regex.match('^(/[^/ ]*)+/?$', document.text)
+        if not ok:
+            raise ValidationError(
+                message='Please enter a valid pathname',
+                cursor_position=len(document.text))  # Move cursor to end
+
+
+class Message:
+    def __init__(self, topic, json_msg):
+        self.json = json_msg
+        self.topic = topic
+
+    def execute(self):
+        # produce to control topic
+        try:
+            producer.produce(self.topic, self.json, callback=delivery_callback)
+        except BufferError:
+            sys.stderr.write('%% Local producer queue is full (%d messages awaiting delivery): try again\n' %
+                             len(producer))
+
+        producer.flush()
+
+
+# clear terminal
+def clear():
+    os.system('cls' if os.name == 'nt' else 'clear')
+
+
+# per message callback
+def delivery_callback(err, msg):
+    if err:
+        sys.stderr.write('%% Message failed delivery: %s\n' % err)
+    else:
+        sys.stderr.write('%% Message delivered to %s [%d] @ %d\n' %
+                         (msg.topic(), msg.partition(), msg.offset()))
+
+
+# interactive command function
+def command_interface():
+    global run, daq_state, daq_error, daq_feedback, daq_evt_rate, daq_events, daq_kB_rate

```

Comment:
These are shared between the threads getting the statuses and waiting for command input. Both containing loops that run until program exit. I can't think of another way to share the variables.

Reply:
I see I can share these variables across the threads. Will attempt that next. I won't need locks since the one thread is read only and there's no real implication to an out of date message.

Reply:
This is a common paradigm in Python. It's the correct way to handle functions. Most everything in Python is global since it's a script language. I don't see any reason to change where these at the moment

Reply:
> I won't need locks since the one thread is read only and there's no real implication to an out of date message.

Okay, that seems a sensible way of avoiding potential race conditions. While I still don't like global variables, in general, it does appear that you've at least covered your bases in this specific case.